File: algorithms/[1178]猜字谜/solution.ts
// topics = ["位运算", "哈希表"]

// 参考 https://leetcode-cn.com/problems/number-of-valid-words-for-each-puzzle/solution/shou-hua-tu-jie-si-lu-jie-xi-leetcode-11-12dy/
function findNumOfValidWords(words: string[], puzzles: string[]): number[] {
  const map = new Map<number, number>();
  for (const word of words) {
    const bits = getBits(word);
    if (map.has(bits)) {
      map.set(bits, map.get(bits)! + 1);
    } else {
      map.set(bits, 1);
    }
  }

  const len = puzzles.length;
  const res = new Array<number>(len).fill(0);
  for (let i = 0; i < len; i += 1) {
    const puzzle = puzzles[i];
    const puzzleBit = getBits(puzzle);
    const first = getBits(puzzle[0]);

    // 枚举 puzzle 对应二进制数的子集，将其在哈希映射中对应的值累计入答案
    let n = puzzleBit;
    while (n > 0) {
      if ((n & first) != 0 && map.has(n)) {
        res[i] += map.get(n)!;
      }
      n = (n - 1) & puzzleBit;
    }
  }

  return res;
}

// 转二进制数表示
// 单词只包含 26 个小写字母，可以压缩到一个长为 26 的 int 中。int 中的每一位取 0 和 1 表示字符是否出现过。
function getBits(s: string): number {
  let res = 0;
  for (const c of s) {
    const offset = c.charCodeAt(0) - 97;
    const status = 1 << offset;
    res = res | status;
  }
  return res;
}

File: algorithms/[1178]猜字谜/solution.py
# topics = ["位运算", "哈希表"]

from collections import Counter
from typing import Dict, List


class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        # 暴力 超时
        ans: List[int] = []

        for puzzle in puzzles:
            cnt = 0
            pset = set(puzzle)
            first = puzzle[0]
            for wset in [set(word) for word in words]:
                if first in wset and pset & wset == wset:
                    cnt += 1
            ans.append(cnt)

        return ans

    def findNumOfValidWords2(self, words: List[str], puzzles: List[str]) -> List[int]:
        """
        二进制压缩状态
        time O(m|w| + n^(2|p|)), space O(m)
        m 和 n 分别为数组 words 和 puzzles 的长度, |w| 是 word 的最大长度 50, |p| 是 puzzle 的最大长度 7
        """
        frequency: Dict[int, int] = Counter()

        # 计算出每一个 word 对应的二进制数，存入计数字典
        for word in words:
            mask = 0
            for ch in word:
                # 使用一个长度为 26 的二进制数表示
                mask |= 1 << (ord(ch) - ord("a"))

            # 如果二进制数中包含的 1 的数量大于 7，那么它一定无法作为谜底
            if str(bin(mask)).count("1") <= 7:
                frequency[mask] += 1

        ans: List[int] = []
        for puzzle in puzzles:
            total = 0
            mask = 0
            for i in range(1, 7):
                mask |= 1 << (ord(puzzle[i]) - ord("a"))

            # 枚举二进制子集
            subset = mask
            while subset:
                s = subset | (1 << (ord(puzzle[0]) - ord("a")))
                if s in frequency:
                    total += frequency[s]
                subset = (subset - 1) & mask

            # 在枚举子集的过程中，要么会漏掉全集 mask，要么会漏掉空集
            # 这里会漏掉空集，因此需要额外判断空集
            if (1 << (ord(puzzle[0]) - ord("a"))) in frequency:
                total += frequency[1 << (ord(puzzle[0]) - ord("a"))]

            ans.append(total)

        return ans
